grpcbp/client_middlewares_test.go

```diff
@@ -0,0 +1,174 @@
+package grpcbp
+
+import (
+	"context"
+	"encoding/json"
+	"net"
+	"testing"
+	"time"
+
+	pb "github.com/grpc-ecosystem/go-grpc-middleware/testing/testproto"
+	"github.com/opentracing/opentracing-go"
+	"github.com/reddit/baseplate.go/mqsend"
+	"github.com/reddit/baseplate.go/tracing"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/test/bufconn"
+)
+
+func TestMonitorInterceptor(t *testing.T) {
+	// create test server
+	l, _ := initServer(t)
+
+	// create test client with monitoring interceptor
+	conn := initClient(t, l, grpc.WithUnaryInterceptor(
+		MonitorInterceptor(
+			MonitorClientArgs{
+				ServiceSlug: "test",
+			},
+		),
+	))
+
+	// instantiate gRPC client
+	ctx, mmq := initServerSpan(t)
+	client := pb.NewTestServiceClient(conn)
+
+	t.Run("span-success", func(t *testing.T) {
+		_, err := client.Ping(ctx, &pb.PingRequest{})
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// drain recorder to validate spans
+		msg, err := drainRecorder(mmq)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		var trace tracing.ZipkinSpan
+		err = json.Unmarshal(msg, &trace)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		got := trace.Name
+		want := "test.Ping"
+		if got != want {
+			t.Errorf("got %s, want: %s", got, want)
+		}
+
+		for _, annotation := range trace.BinaryAnnotations {
+			if annotation.Key == "error" {
+				t.Error("got error span, want: success span")
+				break
+			}
+		}
+	})
+
+	t.Run("span-error", func(t *testing.T) {
+		_, err := client.PingError(ctx, &pb.PingRequest{})
+		if err == nil {
+			t.Error("got no error, want error")
+		}
+
+		// drain recorder to validate spans
+		msg, err := drainRecorder(mmq)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		var trace tracing.ZipkinSpan
+		err = json.Unmarshal(msg, &trace)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		got := trace.Name
+		want := "test.PingError"
+		if got != want {
+			t.Errorf("got %s, want: %s", got, want)
+		}
+
+		for _, annotation := range trace.BinaryAnnotations {
+			if annotation.Key == "error" {
+				return
+			}
+		}
+		t.Error("got success span, want: error span")
+	})
+}
+
+func initServer(t *testing.T, opts ...grpc.ServerOption) (*bufconn.Listener, *mockService) {
+	t.Helper()
+
+	// create in-memory connection to reproduce network behavior but without
+	// the need to clean up OS-level resources
+	l := bufconn.Listen(1024 * 1024)
+	s := grpc.NewServer(opts...)
+	service := &mockService{}
+	pb.RegisterTestServiceServer(s, service)
+	go func() {
+		if err := s.Serve(l); err != nil {
+			t.Error(err)
+		}
+	}()
+	t.Cleanup(s.Stop)
+
+	return l, service
+}
+
+func initClient(t *testing.T, l *bufconn.Listener, opts ...grpc.DialOption) *grpc.ClientConn {
+	// configure required dial option and let test add any further options
+	bufDialer := func(context.Context, string) (net.Conn, error) {
+		return l.Dial()
+	}
+	opts = append([]grpc.DialOption{
+		grpc.WithContextDialer(bufDialer),
```

Comment:
just `l.DialContext`?

Reply:
`bufconn.Listener` does not provide that method. Speaks for your point to test the real thing :wink: